### ISSUE TITLE
Prior and Posterior sentence generator generates same sentences

### ISSUE THREAD

**Issue by bjayakumar** [jxhe/vae-lagging-encoder#5]
I  trained the model with my own data. My dataset size is 5M. After I trained the model, I ran the reconstruction code.

```python text.py --dataset [dataset] --decode_from [pretrained model path] --decode_input [a text file for reconstruction]```

I end up getting  the same sentences

My config:

 ```

params={
    'enc_type': 'lstm',
    'dec_type': 'lstm',
    'nz': 32,
    'ni': 512,
    'enc_nh': 1024,
    'dec_nh': 1024,
    'dec_dropout_in': 0.5,
    'dec_dropout_out': 0.5,
    'batch_size': 32,
    'epochs': 100,
    'test_nepoch': 5,
    'train_data': 'datasets/bj/train.txt',
    'val_data': 'datasets/bj/valid.txt',
    'test_data': 'datasets/bj/test.txt'
}
```

**Comment by jxhe** [comment]
If you are getting the same sentences no matter what z is, that indicates that the decoder does not condition on z (i.e. posterior collapse). By default greedy decoding is used, thus decoder would generate the same sequence if it is not conditioned on z. 

Can you post the KL value, mutual information, and number of active units during training ? It seems the training suffers from posterior collapse.

**Comment by bjayakumar** [comment]
I did not redirect my std output to a log. Let me redirect, re-train, and get back to you. Sorry

**Comment by jxhe** [comment]
Sorry for the inconvenience, it's my bad, I should have included a logger to save logs into files by default. Will include the logger today.

**Comment by dongqian0206** [comment]
I just change the number of hidden units and the dimension of latent variable. The performance on the YAHOO dataset:
KL=6.65
MI=2.987
AU=17

The generated sentences are almost the same.

**Comment by jxhe** [comment]
Sorry I only tested generation on Yelp dataset and it looked ok. Can you provide the config file you used and running command so that I can reproduce the results ? I am curious about what is happening here.

**Comment by dongqian0206** [comment]
Hi. The running command is the same as what you mentioned.

python text.py --decode_from ./models/yahoo/yahoo_aggressive1_kls010_warm10_0_0_1111.pt --dataset yahoo

For yahoo dataset, I directly use the config file, config_yahoo.py

params = {
    'enc_type': 'lstm',
    'dec_type': 'lstm',
    'nz': 32,
    'ni': 512,
    'enc_nh': 1024,
    'dec_nh': 1024,
    'dec_dropout_in': 0.5,
    'dec_dropout_out': 0.5,
    'batch_size': 32,
    'epochs': 100,
    'test_nepoch': 5,
    'train_data': './datasets/yahoo/yahoo.train.txt',
    'val_data': './datasets/yahoo/yahoo.valid.txt',
    'test_data': './datasets/yahoo/yahoo.test.txt'
}

And sorry, I used the **wrong term**, they are not almost the same. Some examples are as follows.

how do you know if a man is gay ? i think he is gay . he is gay . he is gay . </s>
how do i get rid of _UNK _UNK ? i have a _UNK _UNK and i have a _UNK _UNK . i have tried _UNK and _UNK . i
how do i get a _UNK to join ? i 'm a _UNK , and i 'm looking for a good place to start . i 'm looking for
what is the best way to get a _UNK ? i 'm not sure if you 're going to be a good guy , but you 'll have to
how do i get rid of a _UNK _UNK in my mouth ? i have a _UNK _UNK and i have a _UNK _UNK . i have a _UNK
how do i get a _UNK ? i am a _UNK student and i want to know how to get a job in the _UNK area . i am
how do i get rid of _UNK _UNK in my computer ? i have a _UNK _UNK _UNK _UNK _UNK _UNK _UNK _UNK _UNK _UNK _UNK _UNK _UNK _UNK
how do you say `` _UNK '' ? i 'm not sure what you mean . i 'm not sure what you mean . i 'm not sure what

**Comment by bjayakumar** [comment]
@jxhe 
Log file and plot are here.

https://drive.google.com/drive/u/0/folders/13fVAUKI0GWud0bF312RBTbokfyek4-V3

**Comment by jxhe** [comment]
> @jxhe
> Log file and plot are here.
> 
> https://drive.google.com/drive/u/0/folders/13fVAUKI0GWud0bF312RBTbokfyek4-V3

After reading the logs, the model always generates the same sentence because your training clearly collapses (since KL is 0). One abnormal thing I noticed from your logs is that your validation loss actually increases after the first epoch, this does not happen in my experiments with different datasets.  My suggestion for you is to run the experiments without KL annealing (set `kl_start` as 1.0, note that KL annealing doesn't always help), maybe also change the decay epoch threshold value to a smaller one (for example, change 15 to 10):
https://github.com/jxhe/vae-lagging-encoder/blob/2d2497296604cb79d60a235240c9937425b3bc3c/text.py#L485

If you can share the dataset, I am also happy to help run.

**Comment by jxhe** [comment]
@dongqian0206 
I am re-training the model on the Yahoo dataset right now, will get back to you tomorrow.

**Comment by bjayakumar** [comment]
Thanks for your help. I am unable to share dataset as I work for an employer. I will update you soon with results. Again. much appreciated. 

As a side, my training sample length is very small. The avg length of the sentence is 8 words. Does it matter in posterior collapse?

**Comment by jxhe** [comment]
> Thanks for your help. I am unable to share dataset as I work for an employer. I will update you soon with results. Again. much appreciated.
> 
> As a side, my training sample length is very small. The avg length of the sentence is 8 words. Does it matter in posterior collapse?

I have ran the code on short sentences before (avg length < 10) and it doesn't collapse, thus I think it doesn't matter. Also, your dataset seems pretty large and each run costs a lot of time, I would suggest you try running experiments on a relatively small subset of the data for debug purpose ;)

By the way, on the provided `synthetic` dataset (avg length=10), running the code with KL annealing suffers from posterior collapse but running without KL annealing doesn't.

**Comment by bjayakumar** [comment]
Make sense. I am running with 100,000 training examples. Should I need to reduce further?

**Comment by jxhe** [comment]
100,000 is find, I think each run would finish within 10 hours in this scale.

**Comment by jxhe** [comment]
> Make sense. I am running with 100,000 training examples. Should I need to reduce further?

100,000 is fine, I think each run would finish within 10 hours in this scale.

**Comment by bjayakumar** [comment]
I wanted to update you that your idea works. The posterior reconstruction does not have duplicates of the sentence. However, I see there are only 15 new sentences. Those 15 sentences are duplicated all over the file.

**Comment by jxhe** [comment]
> Hi. The running command is the same as what you mentioned.
> 
> python text.py --decode_from ./models/yahoo/yahoo_aggressive1_kls010_warm10_0_0_1111.pt --dataset yahoo
> 
> For yahoo dataset, I directly use the config file, config_yahoo.py
> 
> params = {
> 'enc_type': 'lstm',
> 'dec_type': 'lstm',
> 'nz': 32,
> 'ni': 512,
> 'enc_nh': 1024,
> 'dec_nh': 1024,
> 'dec_dropout_in': 0.5,
> 'dec_dropout_out': 0.5,
> 'batch_size': 32,
> 'epochs': 100,
> 'test_nepoch': 5,
> 'train_data': './datasets/yahoo/yahoo.train.txt',
> 'val_data': './datasets/yahoo/yahoo.valid.txt',
> 'test_data': './datasets/yahoo/yahoo.test.txt'
> }
> 
> And sorry, I used the **wrong term**, they are not almost the same. Some examples are as follows.
> 
> how do you know if a man is gay ? i think he is gay . he is gay . he is gay . 
> how do i get rid of _UNK _UNK ? i have a _UNK _UNK and i have a _UNK _UNK . i have tried _UNK and _UNK . i
> how do i get a _UNK to join ? i 'm a _UNK , and i 'm looking for a good place to start . i 'm looking for
> what is the best way to get a _UNK ? i 'm not sure if you 're going to be a good guy , but you 'll have to
> how do i get rid of a _UNK _UNK in my mouth ? i have a _UNK _UNK and i have a _UNK _UNK . i have a _UNK
> how do i get a _UNK ? i am a _UNK student and i want to know how to get a job in the _UNK area . i am
> how do i get rid of _UNK _UNK in my computer ? i have a _UNK _UNK _UNK _UNK _UNK _UNK _UNK _UNK _UNK _UNK _UNK _UNK _UNK _UNK
> how do you say `` _UNK '' ? i 'm not sure what you mean . i 'm not sure what you mean . i 'm not sure what

@dongqian0206 

I got similar results when sampling from the model, and I think the problem lies at the decoding strategy. It seems greedy decoding tends to generate a lot of UNK tokens.  It you try sampling or beam search decoding the results would be much more reasonable. You can specify decoding strategy through `--decoding_strategy`

**Comment by jxhe** [comment]
> sentences
> I wanted to update you that your idea works. The posterior reconstruction does not have duplicates of the sentence. However, I see there are only 15 new sentences. Those 15 sentences are duplicated all over the file.

@bjayakumar 

You can try using beam search decoding to see if results make more sense. However, I want to let you know that text reconstruction under a VAE model is a very challenging task, and still an open research problem. First, as you know posterior collapse happens when training VAE on text, and existing solutions from many papers only mitigate this issue to some extent, **NOT** solve it. This means the decoder doesn't condition on latent variable well thus it is hard to reconstruct. Second, there is no attention in VAE, which means the latent variable has to capture ALL information of the sentence to achieve perfect reconstruction -- this is difficult, especially for long sequences.

 I was focusing on text VAE reconstruction in the past months, and I have tried training an autoencoder (without KL term) and it proved difficult for an autoencoder to achieve good reconstruction without attention mechanism (in the case sentence length is reasonably long (30) and latent variable is low-dimensional (32)).

If you are familiar with related literature about posterior collapse, there are few VAE papers doing text reconstruction experiments. The [KL annealing paper](https://arxiv.org/abs/1511.06349) reported several examples about text reconstruction, and from those you can see even after careful tuning the reconstruction is not good on a very simple dataset and very short text sequences.

**Comment by bjayakumar** [comment]
Thanks man. I hope you can direct me. I am working to generate similar sentences for the given sentence.
As a baseline model, I used WordNet to create possible sentences for the given sentence.  I then used LM to pick sentences from the set based on sentence prob.
As a next step, I tried VAE with posterior reconstruction. I do not get good results. I used your code.
Note: All sentences are short (like 9 words).


What other methods I should try to generate better similar sentences?

**Comment by jxhe** [comment]
Sorry I can't help more on this. The task you are working on sounds like sentence rephrasing/paraphrase sentence generation ? I am not familiar with this task, and after taking a look on the web I found that there are already some ready-to-use online tools to rephrase sentences. I feel like their performance is definitely better than a VAE model since VAE is not quite mature on text modeling yet.

**Comment by bjayakumar** [comment]
Thanks
